### ISSUE TITLE
Documentation

### ISSUE THREAD

**Issue by cristianchaparroa** [empirefox/gin-oauth2#1]
Please, could you provide some examples to use it?

**Comment by empirefox** [comment]
Example is [here](https://github.com/empirefox/gin-oauth2/tree/master/example) now.
1. implement goauth.OauthUser
2. create goauth.Config and add provider settings
3. create and use middleware
4. serve providers

Document is not ready yet.

**Comment by cristianchaparroa** [comment]
Sorry, I don't understand how to use it with Facebook?

**Comment by empirefox** [comment]
Example is updated. From previous version, we only need a few lines:

1 add provider to config on api server

``` go
goauthConfig.AddProvider("facebook", "/auth/facebook", os.Getenv("FB_CLIENT_ID"), os.Getenv("FB_CLIENT_SECRET"))
```

2 add provider to config on client

``` typescript
const FACEBOOK_PROVIDER: IOauth2Options = {
  clientId: ClientIds['facebook']
}
```

``` typescript
providers: { mock: MOCK_PROVIDER, facebook: FACEBOOK_PROVIDER }
```

``` html
      <button class="btn btn-block btn-facebook" (click)="authenticate('facebook')">
        <i class="fa fa-facebook"></i> sign in with Facebook
      </button>
```

The client uses [ng2-ui-auth](https://github.com/ronzeidman/ng2-ui-auth).
It's the same when working with [satellizer](https://github.com/sahat/satellizer).

3 Repeat the above 2 steps if want more providers support

4 `implement goauth.OauthUser` is a daunting task, because it's our own buissess code, but there is no need for more changes when adding more providers

**Comment by cristianchaparroa** [comment]
Sorry, I just wanna build the `server side` so
1. What exactly is the parameter `Origin` in the `goauthConfig`
2. is not clear how to use the end points for me
3. One the endpoint that show gin is `/auth/facebook` what data must to send to use it?
4. I can't find the `func` `goauthConfig.GetUser(c)`

**Comment by empirefox** [comment]
Sorry for my unclear! This repo is designed to be used with satellizer. Not sure others are suitable.

1.goauth.Config.Origin
  Say you separete your website into 2 parts:
  (A)www.xxx.com is front site based on angular/react, and you deploy it on a CDN.
  (B)api.xxx.com is buissess server, which control the oauth2(`the server side`).
  Then the Origin should be "https://www.xxx.com", and it will be used as redirect_url. We do not support path now.

2.EndPoint
  EndPoint is predefined in goauth.ProviderPresets with some more properties.
  For now the presets contain facebook, google, github, instagram, linkedin, live, yahoo, baidu.
  But not for "mock", we need add it to presets manually.
  Then use it like

``` go
    goauthConfig.AddProvider("facebook", "/auth/facebook", os.Getenv("FB_CLIENT_ID"), os.Getenv("FB_CLIENT_SECRET"))
```

3.POST /auth/facebook payload
  It is based on satellizer, but not all.

``` js
  {
    code: 'code...',
    redirectUri: 'https://...'
  }
```

  `code` is required.
  `redirectUri` is optional, and it is only used when goauth.Config.Origin+preset.RedirectEnd == "".
  Others are ignored.

4.func goauthConfig.GetUser(c)
  Added that for creating example. In the same time some dependencies are upgrade. Please use `go get -u`

**Comment by cristianchaparroa** [comment]
What is exactly the `code` parameter is the code generated by facebook, really?,  and it works the same for the other providers?

**Comment by empirefox** [comment]
Yes, it is. Though most of the preset providers are not tested, it should work the same for the others.
But other frontend frameworks may not be compatible.
